pkg/apis/core/types.go

```diff
@@ -2825,6 +2826,18 @@ type PodAffinityTerm struct {
 	// An empty selector ({}) matches all namespaces.
 	// +optional
 	NamespaceSelector *metav1.LabelSelector
+	// MatchLabelKeys is a set of pod label keys to select which pods will
```

Comment:
I am wondering about the case where I want to do something like: I want to have exclusive 1:1 job to domain placement, so I want to set pod anti-affinity to exclude all other jobs, but not myself, which requires `NotIn` myself anti-affinity constraint. I don't think we can express such a constraint using MatchLabelKeys since it is limited to `In` semantics.

I wonder if we should encapsulate MatchLabelKeys in another struct that allows specifying an operator, it would be limited to `In` or `NotIn`

```
type MatchLabelSelector {
  MatchLabelKeys []string
  Operator LabelSelectorOperator
}
```

and then in the podspec we have:
```
MatchLabelSelectors MatchLabelSelector[]
```

Reply:
but this starts to look like `LabelSelector`, but without the values :)

Reply:
Cool, it's a good usecase to extend MatchLabelKeys.

Let me make sure I understand your usecase correctly:

Your expected behavior is:
- The scheduler fetches the label value from a Pod by `matchLabelKey`. (You have `MatchLabelKeys` in `MatchLabelSelector`, but I guess it should be `MatchLabelKey` so that we can define an operator for each matchLabelKey.)
- If `operator` is `In`, it just merges `key in (value)` with the existing labelSelector. (the same as MatchLabelKeys of the current implementation)
- If `operator` is `NotIn`, it merges `key notin (value)` with the existing labelSelector.

Then actual example:

I assume you'd like to use it with jobset.
https://github.com/kubernetes-sigs/jobset/blob/main/docs/concepts/README.md#jobset-labels

So, I'm thinking you want to do something like this.

```
spec:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - matchLabelSelector: 
            matchLabelKey: jobset.sigs.k8s.io/jobset-name // or jobset.sigs.k8s.io/replicatedjob-name you mean?
            operator NotIn
```

Then, the scheduler builds labelSelector like the above description, and it's gonna behave "This Pod hates all Pods with the different value in `jobset.sigs.k8s.io/jobset-name`"

Are these my understanding correct?

Reply:
> I want to have exclusive 1:1 job to domain placement, so I want to set pod anti-affinity to exclude all other jobs, but not myself

@ahg-g can this be implemented via a compound podaffinity + pod-antiaffinity?

- podaffinity to co-locate pods belong to the same Deployment/Job
- pod-antiaffinity to stay away pods belong to another Deployment/job

<details><summary>deploy1.yaml</summary>
<p>

apiVersion: apps/v1
kind: Deployment
metadata:
  name: app1
spec:
  replicas: 2
  selector:
    matchLabels:
      app: app1
  template:
    metadata:
      labels:
        app: app1
    spec:
      affinity:
        podAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - app1
            topologyKey: kubernetes.io/hostname
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: NotIn
                values:
                - app1
            topologyKey: kubernetes.io/hostname
      containers:
      - name: pause
        image: k8s.gcr.io/pause:3.7

</p>
</details> 

<details><summary>deploy2.yaml</summary>
<p>

apiVersion: apps/v1
kind: Deployment
metadata:
  name: app2
spec:
  replicas: 2
  selector:
    matchLabels:
      app: app2
  template:
    metadata:
      labels:
        app: app2
    spec:
      affinity:
        podAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - app2
            topologyKey: kubernetes.io/hostname
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: NotIn
                values:
                - app2
            topologyKey: kubernetes.io/hostname
      containers:
      - name: pause
        image: k8s.gcr.io/pause:3.7

</p>
</details> 

```shell
⇒  k get node
NAME     STATUS   ROLES    AGE     VERSION
node-1   Ready    <none>   29m     kwok-v0.2.0
node-2   Ready    <none>   29m     kwok-v0.2.0
node-3   Ready    <none>   2m38s   kwok-v0.2.0
⇒  k apply -f deploy1.yaml
⇒  k apply -f deploy2.yaml
⇒  k get po -o wide
NAME                    READY   STATUS    RESTARTS   AGE   IP         NODE     NOMINATED NODE   READINESS GATES
app1-559475955c-bmq2c   1/1     Running   0          15s   10.0.0.2   node-3   <none>           <none>
app1-559475955c-m6htn   1/1     Running   0          15s   10.0.0.4   node-3   <none>           <none>
app2-5b6d44fd98-hzw8d   1/1     Running   0          2s    10.0.0.5   node-2   <none>           <none>
app2-5b6d44fd98-w4mxg   1/1     Running   0          2s    10.0.0.3   node-2   <none>           <none>
```

Reply:
> Except that the user that writes the JobSet doesn't know the label values of the individual Jobs.

Then can it specify dedicated label for each JobSet, and use `Exists` in podaffinity, and `NotExists` in pod-antiaffinity to achieve the goal?

Reply:
All jobs have the same label key, but with different values.

Reply:
Aldo responded already, but re-iterate, a JobSet allows creating [a group of replicated jobs from a template](https://github.com/kubernetes-sigs/jobset/blob/749a77c2647ac46b309b8227f66293e4116b97ed/api/v1alpha1/jobset_types.go#L90), so the template is shared between all jobs, the user can't change the label per job.

Reply:
@sanposhiho yes, but the case you have can be done manually (we know the jobset name, so we can explicitly the value); the case where it is not possible is doing it across jobs within a replicatedJob:

```
spec:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - matchLabelSelectors: 
          - matchLabelKey: job-name
            operator NotIn
```

The reason the use can't set the value explicitly is because the template is shared between [all the jobs of a ReplicatedJob](https://github.com/kubernetes-sigs/jobset/blob/749a77c2647ac46b309b8227f66293e4116b97ed/api/v1alpha1/jobset_types.go#LL95C1-L95C1)

Reply:
to give a concrete example:

```
affinity:
        podAffinity:  # ensures the pods of this job land on the same rack
          requiredDuringSchedulingIgnoredDuringExecution:
          - matchLabelSelectors:
              - matchLabelKey: job-name
              operator: In
             topologyKey: rack
        podAntiAffinity: # ensures only this job lands on the rack
          requiredDuringSchedulingIgnoredDuringExecution:
          - matchLabelSelectors:
              - matchLabelKey: job-name
              operator: NotIn
          - labelSelector:
              matchExpressions:
              - key: job-name
                operator: Exists
            namespaceSelector: {}
            topologyKey: rack
```

Reply:
Thanks, I'll create an update PR for KEP side. This usecase seems pretty clear to me.

Reply:
PTAL: https://github.com/kubernetes/enhancements/pull/4001

Reply:
I understand your usecase now. Add some comments in @sanposhiho 's KEP update.

Reply:
I think this was resolved with mismatchLabelKeys.  If so, can this thread be resolved?

Reply:
@sanposhiho please resolve(fold) this conversation.

Reply:
Yes. Sorry, too many historical threads here.